pkg/backend/display/progress.go

```diff
@@ -661,99 +668,203 @@ func (display *ProgressDisplay) processEndSteps() {
+						display.opts.Color.Colorize(
+							colors.BrightBlue+columns[typeColumn]+" ("+columns[nameColumn]+"):"+colors.Reset))
 				}
-			}
-		}
-	}
 
-	// If we get stack outputs, display them at the end.
-	var wroteOutputs bool
-	if display.stackUrn != "" && !display.opts.SuppressOutputs {
-		stackStep := display.eventUrnToResourceRow[display.stackUrn].Step()
+				for _, line := range lines {
+					line = strings.TrimRightFunc(line, unicode.IsSpace)
+					display.writeSimpleMessage("    " + line)
+				}
+
+				wrote = true
+			}
 
-		props := engine.GetResourceOutputsPropertiesString(
-			stackStep, 1, display.isPreview, display.opts.Debug,
-			false /* refresh */, display.opts.ShowSameResources)
-		if props != "" {
-			if !wroteDiagnosticHeader {
+			if wrote {
 				display.writeBlankLine()
 			}
+		}
+
+	}
+}
 
-			wroteOutputs = true
-			display.writeSimpleMessage(colors.SpecHeadline + "Outputs:" + colors.Reset)
-			display.writeSimpleMessage(props)
+// printPolicyViolations prints a new "Policy Violoation:" section with all of the violations
+// grouped by policy pack. If no policy violations were encountered, prints nothing.
+func (display *ProgressDisplay) printPolicyViolations() {
+	// Loop through every resource and gather up all policy violations encountered.
+	var policyEvents []engine.PolicyViolationEventPayload
+	for _, row := range display.eventUrnToResourceRow {
+		policyInfo := row.PolicyInfo()
+		if policyInfo == nil || len(policyInfo.PolicyPayloads) == 0 {
+			continue
 		}
+		policyEvents = append(policyEvents, policyInfo.PolicyPayloads...)
+	}
+	if len(policyEvents) == 0 {
+		return
 	}
 
-	// print the summary
-	if display.summaryEventPayload != nil {
-		if !wroteDiagnosticHeader && !wroteOutputs {
+	// Sort policy events by: policy pack name, policy pack version, enforcement level,
+	// policy name, and finally the URN of the resource.
+	sort.SliceStable(policyEvents, func(i, j int) bool {
+		eventI, eventJ := policyEvents[i], policyEvents[j]
+		if packNameCmp := strings.Compare(eventI.PolicyPackName, eventJ.PolicyPackName); packNameCmp != 0 {
+			return packNameCmp < 0
+		}
+		if packVerCmp := strings.Compare(eventI.PolicyPackVersion, eventJ.PolicyPackVersion); packVerCmp != 0 {
+			return packVerCmp < 0
+		}
+		eventIEnforcement := string(eventI.EnforcementLevel)
+		eventJEnforcement := string(eventJ.EnforcementLevel)
+		if enfLevelCmp := strings.Compare(eventIEnforcement, eventJEnforcement); enfLevelCmp != 0 {
+			return enfLevelCmp < 0
+		}
+		if policyNameCmp := strings.Compare(eventI.PolicyName, eventJ.PolicyName); policyNameCmp != 0 {
+			return policyNameCmp < 0
+		}
+		eventIURN := string(eventI.ResourceURN)
+		eventJURN := string(eventJ.ResourceURN)
+		urnCmp := strings.Compare(eventIURN, eventJURN)
+		return urnCmp < 0
+	})
+
+	// Print every policy violation, printing a new header when necessary.
+	header := display.opts.Color.Colorize(colors.SpecHeadline + "Policy Violations:" + colors.Reset)
```

Comment:
consider inlining.  locals often make me wonder if you're oing to use this more than one place.